src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,202 @@
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <math.h>
+
+#include "platform.h"
+
+#ifdef USE_MAG_IST8310
+
+#include "build/debug.h"
+
+#include "common/axis.h"
+#include "common/maths.h"
+
+#include "drivers/bus.h"
+#include "drivers/bus_i2c.h"
+#include "drivers/bus_i2c_busdev.h"
+#include "drivers/bus_spi.h"
+#include "drivers/exti.h"
+#include "drivers/io.h"
+#include "drivers/light_led.h"
+#include "drivers/nvic.h"
+#include "drivers/sensor.h"
+#include "drivers/time.h"
+
+#include "compass.h"
+#include "compass_ist8310.h"
+
+//#define DEBUG_MAG_DATA_READY_INTERRUPT
+
+#define IST8310_MAG_I2C_ADDRESS 0x0C
+
+/* ist8310 Slave Address Select : default address 0x0C
+ *        CAD1  |  CAD0   |  Address
+ *    ------------------------------
+ *         VSS   |   VSS  |  0CH
+ *         VSS   |   VDD  |  0DH
+ *         VDD   |   VSS  |  0EH
+ *         VDD   |   VDD  |  0FH
+ * if CAD1 and CAD0 are floating, I2C address will be 0EH
+ *
+ *
+ * CTRL_REGA: Control Register 1
+ * Read Write
+ * Default value: 0x0A
+ * 7:4  0   Reserved.
+ * 3:0  DO2-DO0: Operating mode setting
+ *        DO3  |  DO2 |  DO1 |  DO0 |   mode
+ *    ------------------------------------------------------
+ *         0   |   0  |  0   |  0   |   Stand-By mode
+ *         0   |   0  |  0   |  1   |   Single measurement mode
+ *                                       Others: Reserved
+ *
+ * CTRL_REGB: Control Register 2
+ * Read Write
+ * Default value: 0x0B
+ * 7:4  0   Reserved.
+ * 3    DREN : Data ready enable control:
+ *      0: disable
+ *      1: enable
+ * 2    DRP: DRDY pin polarity control
+ *      0: active low
+ *      1: active high
+ * 1    0   Reserved.
+ * 0    SRST: Soft reset, perform Power On Reset (POR) routine
+ *      0: no action
+ *      1: start immediately POR routine
+ *      This bit will be set to zero after POR routine
+ */
+
+#define IST8310_REG_DATA 0x03
+#define IST8310_REG_WHOAMI 0x00
+
+// I2C Contorl Register
+#define IST8310_REG_CNTRL1 0x0A
+#define IST8310_REG_CNTRL2 0x0B
+#define IST8310_REG_AVERAGE 0x41
+#define IST8310_REG_PDCNTL 0x42
+
+// Parameter
+// ODR = Output Data Rate, we use single measure mode for getting more data.
+#define IST8310_ODR_SINGLE 0x01
+#define IST8310_ODR_10_HZ 0x03
+#define IST8310_ODR_20_HZ 0x05
+#define IST8310_ODR_50_HZ 0x07
+#define IST8310_ODR_100_HZ 0x06
+
+// Device ID (ist8310 -> 0x10)
+#define IST8310_CHIP_ID 0x10
+#define IST8310_AVG_16  0x24
+#define IST8310_PULSE_DURATION_NORMAL 0xC0
+
+#define IST8310_CNTRL2_RESET 0x01
+#define IST8310_CNTRL2_DRPOL 0x04
+#define IST8310_CNTRL2_DRENA 0x08
+
+static bool ist8310Init(magDev_t *magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    busDeviceRegister(dev);
+
+    busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_100_HZ);
+    delay(5);
+
+    busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(5);
+
+    busWriteRegister(dev, IST8310_REG_PDCNTL, IST8310_PULSE_DURATION_NORMAL);
+    delay(5);
```

Comment:
This is the only big difference against my old code. I didn't set this register. But according to the datashet is correct:
![image](https://github.com/betaflight/betaflight/assets/2673520/c669f0fe-1f3b-4c34-bd75-2a04e2745419)

Reply:
Seeing that bits 5:0 are reserved, maybe we are overwriting it with zero? If we want to modify this register, I suppose then that we need to read it, modify the 7:6 bits, and write it again.

Reply:
Seems you have a better reference manual. Please share.

![image](https://github.com/betaflight/betaflight/assets/8344830/ba0a6d2b-24a6-45da-ac61-e1cc99372ccf)

Also trying the init code you got from Isentek

Reply:
I've several files that they sent to me some years ago, I can pass them to you without problem (contact with me in discord).

But the info is from the datasheet, is here online too: https://intofpv.com/attachment.php?aid=8104